FilmoRate v1.3

NatsCracker opened the pull request:
- Добавлен ErrorHandler
- Добавлены новые поля в Film и User
- Добавлены интерфейсы UserStorage и FilmStorage и реализованы их имплиментации
- Реализован UserService и FilmService
- Обновлены контроллеры FilmController и UserController

Commit:
FilmoRate v1.3
- Добавлен ErrorHandler
- Добавлены новые поля в Film и User
- Добавлены интерфейсы UserStorage и FilmStorage и реализованы их имплиментации
- Реализован UserService и FilmService
- Обновлены контроллеры FilmController и UserController

Commit:
FilmoRate v1.3
- Добавлен ErrorHandler
- Добавлены новые поля в Film и User
- Добавлены интерфейсы UserStorage и FilmStorage и реализованы их имплиментации
- Реализован UserService и FilmService
- Обновлены контроллеры FilmController и UserController

Commit:
Merge remote-tracking branch 'origin/add-friends-likes' into add-friends-likes

# Conflicts:
#	pom.xml
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/UserStorage.java

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,43 +2,55 @@
 
-    private final Map<Long, Film> films = new HashMap<>();
-    private long nextId = 1;
+    public FilmController(FilmService filmService, FilmStorage filmStorage) {
```
filmStorage для чего передается?

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -3,46 +3,59 @@
 
-    private final Map<Long, User> users = new HashMap<>();
-    private long nextId = 1L;
+    public UserController(UserService userService) {
```
Можно использовать ломбок

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,35 @@
+    /* ---------- 400 BAD REQUEST ---------- */
+    @ExceptionHandler({MethodArgumentNotValidException.class, ConstraintViolationException.class, jakarta.validation.ValidationException.class})
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleValidation(Exception e) {
```
Все исключения лучше и логировать

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,77 @@
+        if (!film.getLikeUsers().contains(userId)) {
+            throw new NotFoundException("Фильм с id=" + filmId + " не содержит лайк от пользователя с id=" + userId);
+        }
+        film.getLikeUsers().remove(userId);
```
Метод добавления/удаления лайков лучше добавить в Film

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,77 @@
+
+    // Получение самых популярных фильмов
+    public List<Film> getMostPopular(int count) {
+        int limit = count > 0 ? count : DEFAULT_COUNT;
```
Это лучше добавить в контроллер

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,77 @@
+
+    // Получение фильмов пользователя
+    private Film getFilm(long filmId) {
+        return filmStorage.getFilmById(filmId).orElseThrow(() -> new NotFoundException("Фильм с id=" + filmId + " не найден"));
```
Можно использовать String.formatted вместо явной конкатенации

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,79 @@
+        User friend = getUser(friendId);
+
+        if (!user.getListFriends().contains(friendId)) {
+            user.getListFriends().add(friendId);
```
Метод добавления/удаления друзей можно добавить в User

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,64 @@
+
+    // Нормализация имени пользователя
+    private void normalizaName(User user) {
+        if (user.getName() == null || user.getName().isBlank()) {
```
Это логика сервиса

Commit:
FilmoRate v1.31
- Испралены ошибки;
- Испралена логика;

Commit:
FilmoRate v1.32
- Испралены ошибки;
- Испралена логика;

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,43 +2,54 @@
 
-    private final Map<Long, Film> films = new HashMap<>();
-    private long nextId = 1;
+    public FilmController(FilmService filmService) {
```
Можно использовать ломбок.
Аналогично для остальных классов

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,35 @@
+    /* ---------- 400 BAD REQUEST ---------- */
+    @ExceptionHandler({MethodArgumentNotValidException.class, ConstraintViolationException.class, jakarta.validation.ValidationException.class})
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleValidation(Exception e) {
```
Логирование не добавлено

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,80 @@
+
+    // Добавление лайка к фильму
+    public void addLike(long filmId, long userId) {
+        Film film = Optional.ofNullable(filmStorage.getFilmById(filmId))
```
Лучше сделать так, чтобы репозитории возвращали Optional там, где может вернуться null. Аналогично для остальных сервисов

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,88 @@
+
+    // Добавить друга
+    public void addFriend(long userId, long friendId) {
+        if (userId == friendId) {
```
Эти проверки лучше вынести в контроллер. Аналогично для остальных методов

NikolaiDem commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,57 @@
+    // Добавление пользователя в хранилище
+    @Override
+    public User createUser(User user) {
+        if (user.getName() == null || user.getName().isEmpty()) {
```
Это логика сервиса

Commit:
FilmoRate v1.33
- Добавлено логирование ошибок
- Логика проверок перенесена из сервисов в контроллеры

Commit:
FilmoRate v1.33
- Добавлено логирование ошибок
- Логика проверок перенесена из сервисов в контроллеры

Commit:
Merge remote-tracking branch 'origin/add-friends-likes' into add-friends-likes

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/UserService.java

Commit:
FilmoRate v1.33
- Добавлено логирование ошибок
- Логика проверок перенесена из сервисов в контроллеры
